src/main/scala/dev/kamu/cli/commands/PullCommand.scala

```diff
@@ -20,15 +20,44 @@ class PullCommand(
   private val logger = LogManager.getLogger(getClass.getName)
 
   def run(): Unit = {
-    val datasets =
-      if (all)
-        metadataRepository.getAllDatasets()
-      else
-        datasetIDs
-          .map(DatasetID)
-          .map(metadataRepository.getDataset)
-
-    val numUpdated = datasets
+    val s = metadataRepository.getAllDatasetIDs()
+    val datasets: IndexedSeq[DatasetID] = s.toIndexedSeq
+    val datasetIntID = s.zipWithIndex.toMap
+    val idToInt: DatasetID => Int = d => datasetIntID(d)
+    val intToId: Int => DatasetID = d => datasets(d)
+    val idToDataset: DatasetID => Dataset = d => metadataRepository.getDataset(d)
+
+    val scheduled: Array[Boolean] = new Array(datasets.length)
+    for (i <- scheduled.indices) scheduled.update(i, false)
+    var toSchedule: List[Int] = Nil
+    var plan: Seq[Int] = Nil
+    def isScheduled(d: Int): Boolean = scheduled(d)
+
+    if (all) toSchedule = datasets.indices.toList.reverse
+    else toSchedule = datasetIDs.map(DatasetID).toList.map(idToInt)
+
+    logger.info(s"Your pull request:")
+    for (d <- toSchedule) logger.info(s"  ${intToId(d).toString}")
+
+    while (Nil != toSchedule) {
```

Comment:
This is a whole lot of code for one function, so we should modularize it - extract / break it into smaller cohesive functions.

Few things to think about when doing this:
* Maintain a consistent level of granularity in functions (e.g. mixing low-level sorting algorithms with high level logic like spawning docker containers is no bueno)
* Is this class the right place for this logic at all? I'm sure we'll need it in other commands in future